File: cpp/tests/test_discrete.cpp
#include <cmath>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/search/astar.hpp"
#include "navigation/global_planning/search/bfs.hpp"
#include "navigation/global_planning/search/dijkstra.hpp"
#include "navigation/global_planning/search/theta_star.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Cell;

namespace {

core::ParamSet cfg(const std::string& algo) {
  return core::ParamSet::from_yaml(test::repo_path("configs/global_planning/" + algo + ".yaml"));
}

bool path_is_connected(maps::OccupancyGrid2D& g, const std::vector<Cell>& path) {
  for (size_t i = 0; i + 1 < path.size(); ++i) {
    bool ok = false;
    for (const auto& [v, w] : g.neighbors(path[i])) {
      (void)w;
      if (v == path[i + 1]) ok = true;
    }
    if (!ok) return false;
  }
  return true;
}

// Every reconstructed Theta* edge must be a legal straight move — validated via
// line_of_sight (paths are sparse, not neighbor-adjacent).
bool path_los_clear(maps::OccupancyGrid2D& g, const std::vector<Cell>& path) {
  for (size_t i = 0; i + 1 < path.size(); ++i) {
    if (!g.line_of_sight(path[i], path[i + 1])) return false;
  }
  return true;
}

}  // namespace

// (a) valid/optimal path on a known map --------------------------------------

TEST(Discrete, DijkstraAndAstarAreOptimal) {
  auto g = test::make_grid({"...", "...", "..."});  // open 3x3, 8-connected
  Cell start{2, 0}, goal{0, 2};
  double optimal = 2.0 * std::sqrt(2.0);

  global_planning::DijkstraPlanner dij(cfg("dijkstra"));
  auto rd = dij.plan(g, start, goal, nullptr);
  ASSERT_TRUE(rd.success);
  EXPECT_NEAR(rd.cost, optimal, 1e-9);
  EXPECT_EQ(rd.path.front(), start);
  EXPECT_EQ(rd.path.back(), goal);

  global_planning::AstarPlanner astar(cfg("astar"));
  auto ra = astar.plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_NEAR(ra.cost, optimal, 1e-9);
}

TEST(Discrete, BfsMinimisesMoveCount) {
  auto g = test::make_grid({"...", "...", "..."});
  global_planning::BfsPlanner bfs(cfg("bfs"));
  auto r = bfs.plan(g, Cell{2, 0}, Cell{0, 2}, nullptr);
  ASSERT_TRUE(r.success);
  EXPECT_EQ(r.path.size(), 3u);  // 2 diagonal moves = 3 cells
  EXPECT_TRUE(path_is_connected(g, r.path));
}

// Theta* any-angle path (path 2): open grid, goal directly visible from start, so
// the straight segment beats A*'s grid-locked octile path on a non-diagonal offset.
TEST(Discrete, ThetaStarTakesAnyAngleShortcut) {
  auto g = test::make_grid({"...", "...", "..."});  // open 3x3, 8-connected
  Cell start{2, 0}, goal{0, 1};
  global_planning::ThetaStarPlanner theta(cfg("theta_star"));
  auto rt = theta.plan(g, start, goal, nullptr);
  ASSERT_TRUE(rt.success);
  EXPECT_EQ(rt.path.front(), start);
  EXPECT_EQ(rt.path.back(), goal);
  EXPECT_NEAR(rt.cost, std::sqrt(5.0), 1e-9);
  EXPECT_TRUE(path_los_clear(g, rt.path));
  global_planning::AstarPlanner astar(cfg("astar"));
  EXPECT_LT(rt.cost, astar.plan(g, start, goal, nullptr).cost);
}

// Theta* path 1: a blocker hides the goal (no direct LOS), forcing a turn. The
// any-angle path keeps an interior waypoint, every leg is LOS-clear, and it still
// beats grid-locked A*.
TEST(Discrete, ThetaStarBendsAroundObstacle) {
  auto g = test::make_grid({".....", ".....", "..#..", "..#..", "....."});
  Cell start{4, 0}, goal{0, 4};
  global_planning::ThetaStarPlanner theta(cfg("theta_star"));
  auto rt = theta.plan(g, start, goal, nullptr);
  ASSERT_TRUE(rt.success);
  EXPECT_EQ(rt.path.front(), start);
  EXPECT_EQ(rt.path.back(), goal);
  EXPECT_GE(rt.path.size(), 3u);  // bends -> at least one interior waypoint
  EXPECT_FALSE(g.line_of_sight(start, goal));  // goal genuinely hidden
  EXPECT_TRUE(path_los_clear(g, rt.path));
  global_planning::AstarPlanner astar(cfg("astar"));
  EXPECT_LT(rt.cost, astar.plan(g, start, goal, nullptr).cost);
}

// (b) no-path case ------------------------------------------------------------

TEST(Discrete, NoPathWhenWalledOff) {
  auto g = test::make_grid({".#.", ".#.", ".#."});  // middle column blocks all crossing
  Cell start{2, 0}, goal{0, 2};
  global_planning::AstarPlanner astar(cfg("astar"));
  auto r = astar.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
  global_planning::DijkstraPlanner dij(cfg("dijkstra"));
  EXPECT_FALSE(dij.plan(g, start, goal, nullptr).success);
  global_planning::BfsPlanner bfs(cfg("bfs"));
  auto rb = bfs.plan(g, start, goal, nullptr);
  EXPECT_FALSE(rb.success);
  EXPECT_TRUE(rb.path.empty());
  EXPECT_EQ(rb.cost, 0.0);
  global_planning::ThetaStarPlanner theta(cfg("theta_star"));
  auto rth = theta.plan(g, start, goal, nullptr);
  EXPECT_FALSE(rth.success);
  EXPECT_TRUE(rth.path.empty());
  EXPECT_EQ(rth.cost, 0.0);
}

// (c) param validation failure -----------------------------------------------

TEST(Discrete, BadAstarParamThrows) {
  std::string bad = test::write_temp(
      "astar.yaml",
      "algorithm: astar\ncategory: global_planning\nparams:\n"
      "  - name: heuristic_weight\n    type: float\n    default: 0.5\n    min: 1.0\n    max: 5.0\n"
      "    description: below min\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

TEST(Discrete, BadThetaStarParamThrows) {
  std::string bad = test::write_temp(
      "theta_star.yaml",
      "algorithm: theta_star\ncategory: global_planning\nparams:\n"
      "  - name: heuristic_weight\n    type: float\n    default: 0.5\n    min: 1.0\n    max: 5.0\n"
      "    description: below min\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

File: python/tests/test_discrete.py
"""Discrete planners: optimality / validity, and the no-path case."""

from __future__ import annotations

import math
from pathlib import Path

import pytest
from conftest import config, grid_from, open_grid, write_config

from navigation.core.params import ParamError, ParamSet
from navigation.core.types import Cell
from navigation.global_planning import BFS, AStar, Dijkstra, ThetaStar

_OPTIMAL_DIAG = 4 * math.sqrt(2)  # (4,0)->(0,4) on an open 8-connected grid


def _path_is_connected(grid: object, path: list[Cell]) -> bool:
    for a, b in zip(path, path[1:], strict=False):
        succ = {cell for cell, _ in grid.neighbors(a)}  # type: ignore[attr-defined]
        if b not in succ:
            return False
    return True


def _path_los_clear(grid: object, path: list[Cell]) -> bool:
    # Every reconstructed Theta* edge must be a legal straight move — validate via
    # line_of_sight, not neighbor adjacency, since any-angle paths are sparse.
    for a, b in zip(path, path[1:], strict=False):
        if not grid.line_of_sight(a, b):  # type: ignore[attr-defined]
            return False
    return True


def test_dijkstra_and_astar_are_optimal() -> None:
    grid = open_grid(5, 5)
    dj = Dijkstra(config("dijkstra")).plan(grid, (4, 0), (0, 4))
    ast = AStar(config("astar")).plan(grid, (4, 0), (0, 4))
    assert dj.success and ast.success
    assert dj.cost == pytest.approx(_OPTIMAL_DIAG)
    assert ast.cost == pytest.approx(_OPTIMAL_DIAG)
    assert ast.cost == pytest.approx(dj.cost)


def test_astar_expands_no_more_than_dijkstra() -> None:
    # The heuristic must not make A* explore more than uninformed Dijkstra here.
    grid = open_grid(9, 9)
    dj = Dijkstra(config("dijkstra")).plan(grid, (8, 0), (0, 8))
    ast = AStar(config("astar")).plan(grid, (8, 0), (0, 8))
    assert ast.stats.expanded_nodes <= dj.stats.expanded_nodes


def test_bfs_finds_fewest_edge_path() -> None:
    grid = open_grid(5, 5)
    res = BFS(config("bfs")).plan(grid, (4, 0), (0, 4))
    assert res.success
    # 4 diagonal moves -> 5 waypoints; BFS minimizes edge count.
    assert len(res.path) == 5
    assert _path_is_connected(grid, res.path)


def test_no_path_when_walled_off() -> None:
    # A fully occupied middle column separates start from goal.
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."])
    for planner in (
        BFS(config("bfs")),
        Dijkstra(config("dijkstra")),
        AStar(config("astar")),
        ThetaStar(config("theta_star")),
    ):
        res = planner.plan(grid, (0, 0), (0, 4))
        assert not res.success
        assert res.path == []
        assert res.cost == 0.0


def test_theta_star_takes_any_angle_shortcut() -> None:
    # Path 2: on an open grid the goal is directly visible from start, so Theta*
    # returns the single straight segment (cost = Euclidean), strictly shorter
    # than A*'s grid-locked octile path over the same non-diagonal offset.
    grid = open_grid(3, 3)
    start, goal = (2, 0), (0, 1)
    res = ThetaStar(config("theta_star")).plan(grid, start, goal)
    assert res.success
    assert res.path[0] == start and res.path[-1] == goal
    assert res.cost == pytest.approx(math.hypot(2, 1))
    assert _path_los_clear(grid, res.path)
    ast = AStar(config("astar")).plan(grid, start, goal)
    assert res.cost < ast.cost


def test_theta_star_bends_around_obstacle() -> None:
    # Path 1: a blocker hides the goal from start (no direct LOS), forcing a turn.
    # The any-angle path keeps an interior waypoint near the corner, every leg is
    # LOS-clear, and it still beats the grid-locked A* cost.
    grid = grid_from([".....", ".....", "..#..", "..#..", "....."])
    start, goal = (4, 0), (0, 4)
    res = ThetaStar(config("theta_star")).plan(grid, start, goal)
    assert res.success
    assert res.path[0] == start and res.path[-1] == goal
    assert len(res.path) >= 3  # bends -> at least one interior waypoint
    assert not grid.line_of_sight(start, goal)  # goal genuinely hidden
    assert _path_los_clear(grid, res.path)
    ast = AStar(config("astar")).plan(grid, start, goal)
    assert res.cost < ast.cost


def test_theta_star_rejects_out_of_range_weight(tmp_path: Path) -> None:
    # theta_star declares heuristic_weight >= 1.0; a below-min default must fail
    # at load time, so an invalid weight can never reach plan().
    cfg = write_config(
        tmp_path / "bad_theta.yaml",
        "theta_star",
        [{"name": "heuristic_weight", "type": "float", "default": 0.5,
          "min": 1.0, "max": 5.0, "description": "below min"}],
    )
    with pytest.raises(ParamError):
        ParamSet.from_yaml(cfg)
